[AutoDiff] Closure specialization: specialize branch tracing enums

kovdan01 opened the pull request:
**Re-opened as #85757**

This patch contains part of the changes intended to resolve #68944.

1. Closure info gathering logic.
2. Branch tracing enum specialization logic.
3. Specialization of branch tracing enum basic block arguments in VJP.
4. Specialization of branch tracing enum payload basic block arguments in pullback.

kovdan01 commented:
Tagging @JaapWijnen

kovdan01 commented:
@swift-ci please test

eeckstein commented:
@kovdan01 Sorry that my PR https://github.com/swiftlang/swift/pull/84704 interferes with your work here. It's a big simplification of the pass and it should also make your work easier. If you have some questions about the new pass design, just let me know.

Also, can you add a simple SIL example in a comment describing what this PR is about?

kovdan01 committed:
[AutoDiff] Closure specialization: specialize branch tracing enums

This patch contains part of the changes intended to resolve #68944.

1. Closure info gathering logic.
2. Branch tracing enum specialization logic.
3. Specialization of branch tracing enum basic block arguments in VJP.
4. Specialization of branch tracing enum payload basic block arguments in pullback.
